Use weight to balance Fresh-Water flux

jm-c opened the pull request:
## What changes does this PR introduce?
Allow to correct non-uniformly for global-mean surf.flux imbalance.
- for now, only apply to surface fresh-water flux.
- imbalance is corrected proportionally to 2-D weight field, fixed in time (only read-in once ; add new input file name for weight).
- new (integer) parameter "selectBalanceEmPmR" to replace logical param "balanceEmPmR".

## What is the current behaviour? 
The only way to balance fresh-water flux is to apply uniform correction.

## What is the new behaviour 
new option (for now, only for Fresh-water flux) to correct non-uniformly for (global) imbalance of surface-flux. proportionally to a provided 2-d weight field.

## Does this PR introduce a breaking change? 
- backward compatible new selector "selectBalanceEmPmR" to chose how to balance Frsh-water, still recognized previous logical switch "balanceEmPmR".
- new 2-D weight field is within #ifdef ALLOW_BALANCE_FLUXES, so no additional memory requirement if/when left undefined.

## Other information:
- few issues to discuss (e.g., do we want to implement same alternative correction for net heat flux ? and would it be using same weight ?) that might involve renaming few new params/var: consensus is that this is not necessary, so just keep this new option for net surface fresh-water flux.
-  balanceEmPmR has been removed from pkg/seaice & pkg/thsice and replace with selectBalanceEmPmR=1.

## Suggested addition to `tag-index`
o model:
  - add option to apply a correction for net global fresh-water flux imbalance
    proportional to, fixed in time, 2-D weight field that is read-in from file ;
  - new integer selector "selectBalanceEmPmR" (=0: off, =1: uniform correction,
    =2: correction scaled by weight) that replaces logical switch "balanceEmPmR"
    with backward compatibility setting.

mjlosch commented:
The "balanceEmPmR" scheme has always been a hack when inconsistent forcing would lead to a net fw-flux. The corresponding "balanceQnet" was never really clear to me, because in long simulations the system would, with reasonably bad forcing, reach an equilibrium as surface temperatures adjust to the surface heat fluxes. So I would argue that we would not need to increase the complexity of "balanceQnet".

For "balanceEmPmR", allowing for weighted corrections make sense to me (in the spirit of "flux correction" to reduce local salinity biases). What we would really need, though, is a scheme that collects the fw-imbalance and applies this as a correction in the next forcing cycle (e.g. year, decade, century), but that's not the goad of the PR, I guess.

jrscott commented:
I agree in principle with @mjlosch above.

I have used the balanceEmPmR scheme in various setups, and having the weighting function being available would have been very useful (as opposed to using some offline scheme).
Where you dump E-P is very important obviously in the global system, and often one finds themselves way out of balance.

But I agree the balanceQnet addition is of much less utility; perhaps in some idealized setup applying such balance here might be useful, but in such idealization using a weighting map might be less important, and moreover, less physically obvious what that weighting should be.

jm-c commented:
@mjlosch and @jrscott  thanks for your comments, this is useful. I will do minor adjustment but keep this option just for fresh-water flux, and finish the rest.

jm-c commented:
I finished the little adjustments so I think it's ready fro review.

mjlosch reviewed (approved):
LGTM

jrscott commented:
@jm-c I don't mean to be a pain, everything looks ok except remove_mean.F

- it is a bit confusing in REMOVE_MEAN_RS, but I think this is because this routine is multipurpose, and written this way to handle these other calls from other places. It is just not great doc-wise when routines like this do not have explanatory comments why they are seemingly written in such strange (non-obvious) manner (using parm switches to do different things), if we expect anyone else to have any hope of following the code.
- In the end I don't really understand what you mean by "weighted correction". It seems it simply take the imbalance and multiply by the local(x,y) weight. Why would you use the weights in computing the imbalance? (perhaps I don't understand)

The main idea here is allowing for some non-uniform, physical way to redistribute an imbalance in EmPmR. As noted, not following what you are attempting here... but I'd also note what is done in thsice: the imbalance is weighted by precipitation, which makes a lot of physical sense to me (uniform does not in a global setup). Would that be possible here instead (and skip the file input)? ie why take different approach here vs. what's done in thsice?

(as a further comment, it would be very useful to allow this work with pkg seaice, if we are ever to push this for use in a global config)

jrscott commented:
I'd also go back to @mjlosch comment: "What we would really need, though, is a scheme that collects the fw-imbalance and applies this as a correction in the next forcing cycle (e.g. year, decade, century)". Option to balance weight by P+R or E (let user pick) would be ideal. And have it work consistently with no seaice, seaice, or thsice. This is beyond the scope of this PR obviously, but I think we should be planning toward such goal and raise the lacking functionality here as an issue.

jm-c commented:
@jrscott In reversed order:  If you feel that it's worth to open an issue, please do.

jm-c commented:
@jrscott continuing in reverse order:
Regarding this:
> The main idea here is allowing for some non-uniform, physical way to redistribute an imbalance in EmPmR
I don't believe it's more "physical", just that it's convenient to have this option.
Now I think this set of changes do what this PR describes:
> imbalance is corrected proportionally to 2-D weight field, fixed in time (only read-in once ; add new input file name for weight).
I do think that this option could be useful (and in fact, I know someone who use it); and I thought, from you earlier comment from Jan 11, that you did not disagree. 
And yes, it does not do other things (like weighting by precip, BTW, precip is not a "model" field but only available in some pkgs).

jm-c commented:
@jrscott Continuing. Regarding this:
> (as a further comment, it would be very useful to allow this work with pkg seaice, if we are ever to push this for use in a global config)
I am confused regarding the usefulness of this feature: If it's not really useful without pkg/seaice (because you would prefer to use precip as weight), then it's probably not that useful neither with pkg/seaice.

jm-c commented:
@jrscott Last comment, regarding this:
> it is a bit confusing in REMOVE_MEAN_RS, but I think this is because this routine is multipurpose, and written this way to handle these other calls from other places. It is just not great doc-wise when routines like this do not have explanatory comments why they are seemingly written in such strange (non-obvious) manner (using parm switches to do different things), if we expect anyone else to have any hope of following the code.

REMOVE_MEAN_RS does correct for global-mean imbalance, so that after applying this call the field has zero global mean. In this sense, it does "remove the global mean". It does is uniformly (original option) or proportional to a weight (new option). Not sure that we would gain a lot by having 2 subroutines instead of a single one. Is it what you suggest ?
Also I am not sure that I see clearly what you mean by "written in such strange (non-obvious) manner"; could you be more specific ?

jm-c commented:
@jrscott Thanks for the clarification and your addition to the docs. Here is what I would like to do:

1. review to clarify where "weighted-correction" is used, may be adding "scale by weight" or something like that. But I suggest to keep calling "weight" the - fixed in time - scaling factor "weight2BalanceFlx" and "wghtBalanceFile".
2. document argument list of S/R REMOVE_MEAN_RS, with the 2 methods. I will also add a comment about this S/R being currently only used for 2-D fields. Will see also if I can clarify some pieces of code there.
3. once we are happy with the edits in REMOVE_MEAN_RS I could make same changes to the _RL version. This way it will be cleaner, I think.

jrscott commented:
@jm-c on the first point, agree the current variable names are ok. I was ok with my change to the users manual, ie though the change in the parm table where I added the word "scaled" was sufficient, so not sure you need to do anything further on this point, other than review.

jm-c commented:
I changed a bit the suggested addition to tag-index and added some documentation (S/R arguments) for S/R REMOVE_MEAN_RS. @jrscott and @mjlosch if this is good then will carry the new option to _RL version.

mjlosch commented on model/src/remove_mean.F:
```diff
@@ -113,11 +113,14 @@ SUBROUTINE REMOVE_MEAN_RS(
-C     |   substract it from the same array                       |
-C     *==========================================================*
+C     Correct for global-mean imbalance of global "_RS" array "arrFld"
+C      (i.e., output arrFld has zero mean):
```
new indentation style for comments (o:

jm-c replied:
@mjlosh I am not sure I understand what you mean here.

mjlosch commented on model/src/remove_mean.F:
```diff
@@ -113,11 +113,14 @@ SUBROUTINE REMOVE_MEAN_RS(
+C      or (if arrDr(1) < 0) a correction scaled by "arrhFac" weight.
+C     The global mean imbalance is computed as area (arrArea), mask
+C      (arrMask) and thickness (arrDr) weighted, as well as thickness
+C      factor (arrhFac) weighted (1rst case) or without (2nd case).
```
I may be overstepping but I would write:
`factor (arrhFac) weighted (1rst case) or not (2nd case).`
OR
`factor (arrhFac) weighted (1rst case) or without weighting (2nd case).`

jm-c replied:
I agree, changing this line.

mjlosch reviewed (approved):
I am fine with the recent changes.

jrscott commented:
thx @jm-c additions to subroutine doc are very helpful

jm-c commented:
From my side, this is ready to be merged in (could do this later today). @mjlosch do you agree ?
